AVRO-2079: Add ability to generate Java8 native date/time classes (new)

pvorb opened the pull request:
As there was no progress on #248 for the past few months, I went ahead and resolved the merge conflicts of that PR and also fixed the remarks from @Kuroshii.

Here's a brief overview of what I've changed:

- Reorder imports in TestJava8TimeConversions.java
- Directly cast epochDays to int
- Directly cast millis to int
- Explicitly format java8 and joda time dates in test
- Rename DateTimeLogicalTypeType to DateTimeLogicalTypeImplementation
- Refactor registration of conversions
- Make DateTimeLigicalTypeImplementation field final
- Use parent POM version in Maven plugin tests
- Log warning message about unknown parameter values
- Remove unused imports

I would also suggest to rename the enum value `JAVA8` and all of its uses to `JSR310` because that's more specific to the `java.time.*` APIs introduced in Java 8 (see https://jcp.org/en/jsr/detail?id=310). I can add that change if you agree.

@Kuroshii I would appreciate if you could review this PR again, so maybe we'll have JSR-310 support in the next Avro release. Thanks!

lukejackson commented on lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java:
```diff
@@ -225,6 +268,14 @@ public void setEnableDecimalLogicalType(boolean enableDecimalLogicalType) {
     this.enableDecimalLogicalType = enableDecimalLogicalType;
   }
 
+  public boolean useJodaForDateTimeLogicalTypes() {
```
any reason not to expose the enum here? the two getters are not independent, and the interface essentially exposes four possible states when there are only two.

lukejackson commented on lang/java/maven-plugin/src/test/resources/unit/idl/pom-java8.xml:
```diff
@@ -0,0 +1,69 @@
+    <artifactId>avro-parent</artifactId>
+    <groupId>org.apache.avro</groupId>
+    <version>1.9.0-SNAPSHOT</version>
+    <relativePath>../../../../../../../../../</relativePath>
```
relativePath should point to a specific pom.xml file, rather than a directory, here and in other pom.xml files touched by this change.

lukejackson commented:
+1 on renaming to JSR310. (Note I am not an Avro committer).

lukejackson commented on lang/java/compiler/src/main/velocity/org/apache/avro/compiler/specific/templates/java/classic/record.vm:
```diff
@@ -163,9 +163,9 @@ public class ${this.mangle($schema.getName())}#if ($schema.isError()) extends or
 #if ($this.hasLogicalTypeField($schema))
   protected static final org.apache.avro.Conversions.DecimalConversion DECIMAL_CONVERSION = new org.apache.avro.Conversions.DecimalConversion();
 #if ($this.useJodaForDateTimeLogicalTypes())
```
i think this needs to be refactored to use the new getDateTimeLogicalTypeImplementation() method

pvorb replied:
Good spot!

DolgopolovOleg commented:
Sorry guys, it's really crucial feature for us. What do you think how long will it take to approve it?
Many thanks =)

nandorKollar commented on lang/java/tools/src/main/java/org/apache/avro/tool/SpecificCompilerTool.java:
```diff
@@ -164,7 +181,7 @@ public String getShortDescription() {
     }
 
-    return fileSet.toArray((new File[fileSet.size()]));
+    return fileSet.toArray(new File[0]);
```
Why did you change this? Since we know the size of the result array, the original code (allocating one before toArray, and passing it) is still fine isn't it?

pvorb replied:
[Here is an excellent article](https://shipilev.net/blog/2016/arrays-wisdom-ancients/) that explains why nowadays it's better (despite counter intuitive) to use `toArray(new File[0])` in favor of `toArray(new File[fileSet.size()])`. The old code was not wrong, but I wanted to remove those redundant parentheses and switched to `new File[0]` because I was already editing that line. ;-)

pvorb replied:
@nandorKollar If you prefer `toArray(new File[fileSet.size()])`, I can revert it of course.

pvorb commented:
Recent changes:

- added a commit that replaces `JAVA8` by `JSR310` as suggested by me earlier in this thread. If anybody from the Avro project disagrees with that change, I will remove it.

- removed the runtime dependency on Guava's `Strings` class from the `AbstractAvroMojo`, which caused the archetype modules to fail.

- added a new entry in the `CHANGES.txt` describing the feature

@nandorKollar Is there anything else I can do?

pvorb commented:
@nandorKollar Just saw that you want to remove `CHANGES.txt` in #310, so I removed the new entry in order to prevent a merge conflict.

pvorb commented:
@nandorKollar Is there anything else I can do?

pvorb commented:
@cutting Can you please have a look? Is there anything that needs to be changed before merging this request? It would be nice to get some feedback given the time me and others invested into it.

pvorb commented:
This PR and [its precursor](#248) are by far the most upvoted pull requests for this project. Probably it’s worth considering to merge it?

malcolmrobbins commented:
Hi pvorb,
we're really keen to get hold of Java8 time support within AVRO, even if this means creating our own fork for a while as we're about to undertake a significant project...  What would it take for me to do this ?  i.e. can I pull the changes and apply this to my own fork of the code?  Alternatively is there an AVRO release imminent to save me the trouble?

malcolmrobbins commented:
i.e. can I get hold of vorb:use-java8-time for our "private use" until it becomes generally available

rocketraman commented:
LOL: https://issues.apache.org/jira/browse/AVRO-2210

pvorb commented:
@malcolmrobbins Feel free to do this. Don't expect an official release with this change anytime soon.

malcolmrobbins commented:
Thanks for that - I took a copy of your branch and tried a mvn test install and the build failed on Apache Avro Tools with an error that indicates a dependency on avro-mapred:jar:tests could not be satisfied.  The actual error is:

[ERROR] Failed to execute goal on project avro-tools: Could not resolve dependencies for project org.apache.avro:avro-tools:jar:1.9.0-SNAPSHOT: Could not find artifact org.apache.avro:avro-mapred:jar:tests:1.9.0-SNAPSHOT -> [Help 1]

Do have any pointers as to how I can overcome this?  Thanks in advance.

malcolmrobbins commented:
FYI I overcame this problem.  To do this I simply needed to run the build.sh script in the lang/java folder rather than following the more explicit mvn instructions in the How To Contribute guide.  Now I just need to figure out how to get the avro-maven-plugin to generate Java sources that use Java 8 datetime rather than Joda

kgalieva commented:
@malcolmrobbins You need to add `<dateTimeLogicalTypeImplementation>jsr310</dateTimeLogicalTypeImplementation>` into `configuration` section.

@pvorb thank you! This patch works fine on our schemas!

Fokko commented:
I've ran some tests against [Divolte](https://github.com/divolte/divolte-collector) and it looks great. Can we get this merged?

nandorKollar commented on lang/java/avro/src/main/java/org/apache/avro/data/Jsr310TimeConversions.java:
```diff
@@ -0,0 +1,158 @@
+import java.time.LocalTime;
+import java.util.concurrent.TimeUnit;
+
+public class Jsr310TimeConversions {
```
It would be nice to override getRecommendedSchema methods for each conversion similar as in the Joda time conversion class (TimeConversions)

nandorKollar commented on lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java:
```diff
@@ -153,16 +174,38 @@ public SpecificCompiler(Protocol protocol) {
   }
 
+  /**
+   * Creates a specific compler with the default (Joda) type for date/time related logical types.
```
nit: compler -> compiler

pvorb commented:
@nandorKollar I changed the code according to your requests. Good spot about overriding the `getRecommendedSchema()` methods in each conversion!

nandorKollar commented:
Merged to master, thanks @aukevanleeuwen and @pvorb for working on this!

pvorb commented:
Thanks, @nandorKollar!

Blackdread commented:
Hi, it is merged but avro maven plugin 1.8.0 still does not support that, right?
I am trying to set my avsc file but still get jodatime once generated

Haybu commented:
Any rough estimation when 1.9.0 will be released?

Fokko commented:
Next week I'll cut the branch and release the RC. Please subscribe to the dev maillist so you can help test the RC 👍
